File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/ui/ValidationRunController.java
package net.nemerosa.ontrack.boot.ui;

import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.Selection;
import net.nemerosa.ontrack.model.security.SecurityService;
import net.nemerosa.ontrack.model.structure.*;
import net.nemerosa.ontrack.ui.controller.AbstractResourceController;
import net.nemerosa.ontrack.ui.resource.Pagination;
import net.nemerosa.ontrack.ui.resource.PaginationCountException;
import net.nemerosa.ontrack.ui.resource.PaginationOffsetException;
import net.nemerosa.ontrack.ui.resource.Resources;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.net.URI;
import java.util.List;

import static org.springframework.web.servlet.mvc.method.annotation.MvcUriComponentsBuilder.on;

@RestController
@RequestMapping("/structure")
public class ValidationRunController extends AbstractResourceController {

    private final StructureService structureService;
    private final ValidationRunStatusService validationRunStatusService;
    private final PropertyService propertyService;
    private final SecurityService securityService;

    @Autowired
    public ValidationRunController(StructureService structureService, ValidationRunStatusService validationRunStatusService, PropertyService propertyService, SecurityService securityService) {
        this.structureService = structureService;
        this.validationRunStatusService = validationRunStatusService;
        this.propertyService = propertyService;
        this.securityService = securityService;
    }

    @RequestMapping(value = "builds/{buildId}/validationRuns/view", method = RequestMethod.GET)
    public Resources<ValidationStampRunView> getValidationStampRunViews(@PathVariable ID buildId) {
        // Build
        Build build = structureService.getBuild(buildId);
        // Gets the views
        List<ValidationStampRunView> views = structureService.getValidationStampRunViewsForBuild(build);
        // Converts into a view
        URI uri = uri(on(getClass()).getValidationStampRunViews(buildId));
        return Resources.of(
                views,
                uri
        ).forView(ValidationStampRunView.class);
    }

    @RequestMapping(value = "builds/{buildId}/validationRuns", method = RequestMethod.GET)
    public Resources<ValidationRun> getValidationRuns(@PathVariable ID buildId) {
        return Resources.of(
                structureService.getValidationRunsForBuild(buildId),
                uri(on(ValidationRunController.class).getValidationRuns(buildId))
        ).forView(Build.class);
    }

    @RequestMapping(value = "builds/{buildId}/validationRuns/create", method = RequestMethod.GET)
    public Form newValidationRunForm(@PathVariable ID buildId) {
        Build build = structureService.getBuild(buildId);
        return Form.create()
                .with(
                        Selection.of("validationStampId")
                                .label("Validation stamp")
                                .items(structureService.getValidationStampListForBranch(build.getBranch().getId()))
                )
                .with(
                        Selection.of("validationRunStatusId")
                                .label("Status")
                                .items(validationRunStatusService.getValidationRunStatusRoots())
                )
                .description();
    }

    @RequestMapping(value = "builds/{buildId}/validationRuns/create", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.CREATED)
    public ValidationRun newValidationRun(@PathVariable ID buildId, @RequestBody ValidationRunRequest validationRunRequest) {
        // Gets the build
        Build build = structureService.getBuild(buildId);
        // Gets the validation stamp
        ValidationStamp validationStamp = getValidationStamp(build.getBranch(), validationRunRequest.getValidationStampId(), validationRunRequest.getActualValidationStampName());
        // Gets the validation run status
        ValidationRunStatusID validationRunStatusID = validationRunStatusService.getValidationRunStatus(validationRunRequest.getValidationRunStatusId());
        // Validation run to create
        ValidationRun validationRun = ValidationRun.of(
                build,
                validationStamp,
                0,
                securityService.getCurrentSignature(),
                validationRunStatusID,
                validationRunRequest.getDescription()
        );
        // Creation
        validationRun = structureService.newValidationRun(validationRun);
        // Saves the properties
        for (PropertyCreationRequest propertyCreationRequest : validationRunRequest.getProperties()) {
            propertyService.editProperty(
                    validationRun,
                    propertyCreationRequest.getPropertyTypeName(),
                    propertyCreationRequest.getPropertyData()
            );
        }
        // OK
        return validationRun;
    }

    protected ValidationStamp getValidationStamp(Branch branch, Integer validationStampId, String validationStampName) {
        return structureService.getOrCreateValidationStamp(branch, validationStampId, validationStampName);
    }

    @RequestMapping(value = "validationRuns/{validationRunId}", method = RequestMethod.GET)
    public ValidationRun getValidationRun(@PathVariable ID validationRunId) {
        return structureService.getValidationRun(validationRunId);
    }

    // Validation run status

    @RequestMapping(value = "validationRuns/{validationRunId}/status/change", method = RequestMethod.GET)
    public Form getValidationRunStatusChangeForm(@PathVariable ID validationRunId) {
        ValidationRun validationRun = structureService.getValidationRun(validationRunId);
        return Form.create()
                .with(
                        Selection.of("validationRunStatusId")
                                .label("Status")
                                .items(
                                        validationRunStatusService.getNextValidationRunStatusList(validationRun.getLastStatus().getStatusID().getId())
                                )
                )
                .description()
                ;
    }

    @RequestMapping(value = "validationRuns/{validationRunId}/status/change", method = RequestMethod.POST)
    @ResponseStatus(HttpStatus.CREATED)
    public ValidationRun validationRunStatusChange(@PathVariable ID validationRunId, @RequestBody ValidationRunStatusChangeRequest request) {
        // Gets the current run
        ValidationRun run = structureService.getValidationRun(validationRunId);
        // Gets the new validation run status
        ValidationRunStatus runStatus = ValidationRunStatus.of(
                securityService.getCurrentSignature(),
                validationRunStatusService.getValidationRunStatus(request.getValidationRunStatusId()),
                request.getDescription()
        );
        // Updates the validation run
        return structureService.newValidationRunStatus(run, runStatus);
    }

    /**
     * List of validation runs for a validation stamp
     */
    @RequestMapping(value = "validationStamps/{validationStampId}/validationRuns", method = RequestMethod.GET)
    public Resources<ValidationRun> getValidationRunsForValidationStamp(
            @PathVariable ID validationStampId,
            @RequestParam(required = false, defaultValue = "0") int offset,
            @RequestParam(required = false, defaultValue = "10") int count) {
        // Gets ALL the runs
        List<ValidationRun> runs = structureService.getValidationRunsForValidationStamp(validationStampId, 0, Integer.MAX_VALUE);
        // Total number of runs
        int total = runs.size();
        // Checks the offset and count
        if (offset < 0) {
            throw new PaginationOffsetException(offset);
        } else if (offset > 0 && offset >= total) {
            throw new PaginationOffsetException(offset);
        } else if (count <= 0) {
            throw new PaginationCountException(count);
        }
        // Prepares the resources
        Resources<ValidationRun> resources = Resources.of(
                runs.subList(offset, Math.min(offset + count, runs.size())),
                uri(on(ValidationRunController.class).getValidationRunsForValidationStamp(
                        validationStampId,
                        offset,
                        count
                ))
        );
        // Pagination information
        Pagination pagination = Pagination.of(offset, count, total);
        // Previous page
        if (offset > 0) {
            pagination = pagination.withPrev(
                    uri(on(ValidationRunController.class).getValidationRunsForValidationStamp(
                            validationStampId,
                            Math.max(0, offset - count),
                            count
                    ))
            );
        }
        // Next page
        if (offset + count < total) {
            pagination = pagination.withNext(
                    uri(on(ValidationRunController.class).getValidationRunsForValidationStamp(
                            validationStampId,
                            offset + count,
                            count
                    ))
            );
        }
        // OK
        return resources.withPagination(pagination).forView(ValidationStampRunView.class);
    }

}

File: ontrack-ui/src/main/java/net/nemerosa/ontrack/boot/ui/ValidationRunController.kt
package net.nemerosa.ontrack.boot.ui

import net.nemerosa.ontrack.json.JsonParseException
import net.nemerosa.ontrack.model.exceptions.ValidationRunDataJSONInputException
import net.nemerosa.ontrack.model.security.SecurityService
import net.nemerosa.ontrack.model.structure.*
import net.nemerosa.ontrack.ui.controller.AbstractResourceController
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import kotlin.math.min

@RestController
@RequestMapping("/rest/structure")
class ValidationRunController
@Autowired
constructor(
    private val structureService: StructureService,
    private val validationRunStatusService: ValidationRunStatusService,
    private val validationDataTypeService: ValidationDataTypeService,
    private val securityService: SecurityService
) : AbstractResourceController() {

    @GetMapping("builds/{buildId}/validationRuns/view")
    fun getValidationStampRunViews(@PathVariable buildId: ID): List<ValidationStampRunView> {
        // Build
        val build = structureService.getBuild(buildId)
        // Gets the views
        val views = structureService.getValidationStampRunViewsForBuild(build)
        // Converts into a view
        return views
    }

    @GetMapping("builds/{buildId}/validationRuns")
    fun getValidationRuns(@PathVariable buildId: ID): List<ValidationRun> {
        return structureService.getValidationRunsForBuild(buildId, 0, 100)
    }

    /**
     * Note that some properties, like the `type` of data are provided only when using the DSL
     * or the raw REST API, never through the GUI.
     */
    @PostMapping("builds/{buildId}/validationRuns/create")
    @ResponseStatus(HttpStatus.CREATED)
    fun newValidationRun(
        @PathVariable buildId: ID,
        @RequestBody validationRunRequestForm: ValidationRunRequestForm
    ): ValidationRun {
        // Gets the build
        val build = structureService.getBuild(buildId)
        // Creates the service validation run request from the form
        val validationRunRequest = ValidationRunRequest(
            validationStampName = validationRunRequestForm.actualValidationStampName,
            validationRunStatusId = validationRunRequestForm.validationRunStatusId?.run {
                validationRunStatusService.getValidationRunStatus(this)
            },
            dataTypeId = validationRunRequestForm.validationStampData?.type,
            data = parseValidationRunData(build, validationRunRequestForm),
            description = validationRunRequestForm.description,
            properties = validationRunRequestForm.properties
        )
        // Delegates to the service
        return structureService.newValidationRun(build, validationRunRequest)
    }

    private fun parseValidationRunData(build: Build, validationRunRequestForm: ValidationRunRequestForm): Any? {
        return validationRunRequestForm.validationStampData?.data?.run {
            // Gets the validation stamp
            val validationStamp: ValidationStamp = structureService.getOrCreateValidationStamp(
                build.branch,
                validationRunRequestForm.actualValidationStampName
            )
            // Gets the data type ID if any
            // First, the data type in the request, and if not specified, the type of the validation stamp
            val typeId: String? = validationRunRequestForm.validationStampData.type
                ?: validationStamp.dataType?.descriptor?.id
            // If no type, ignore the data
            return typeId?.run {
                // Gets the actual type
                validationDataTypeService.getValidationDataType<Any, Any>(this)
            }?.run {
                // Parses data from the form
                try {
                    fromForm(validationRunRequestForm.validationStampData.data)
                } catch (ex: JsonParseException) {
                    throw ValidationRunDataJSONInputException(ex, validationRunRequestForm.validationStampData.data)
                }
            }
        }
    }

    @GetMapping("validationRuns/{validationRunId}")
    fun getValidationRun(@PathVariable validationRunId: ID): ValidationRun =
        structureService.getValidationRun(validationRunId)

    // Validation run status

    @PostMapping("validationRuns/{validationRunId}/status/change")
    @ResponseStatus(HttpStatus.CREATED)
    fun validationRunStatusChange(
        @PathVariable validationRunId: ID,
        @RequestBody request: ValidationRunStatusChangeRequest
    ): ValidationRun {
        // Gets the current run
        val run = structureService.getValidationRun(validationRunId)
        // Gets the new validation run status
        val runStatus = ValidationRunStatus(
            ID.NONE,
            securityService.currentSignature,
            validationRunStatusService.getValidationRunStatus(request.validationRunStatusId),
            request.description
        )
        // Updates the validation run
        return structureService.newValidationRunStatus(run, runStatus)
    }

    /**
     * List of validation runs for a validation stamp
     */
    @GetMapping("validationStamps/{validationStampId}/validationRuns")
    fun getValidationRunsForValidationStamp(
        @PathVariable validationStampId: ID,
        @RequestParam(required = false, defaultValue = "0") offset: Int,
        @RequestParam(required = false, defaultValue = "10") count: Int
    ): List<ValidationRun> {
        // Gets ALL the runs
        val runs = structureService.getValidationRunsForValidationStamp(validationStampId, 0, Integer.MAX_VALUE)
        // Prepares the resources
        return runs.subList(offset, min(offset + count, runs.size))
    }

}
